packages/apollo-codegen-swift/src/helpers.ts

```diff
@@ -347,3 +368,52 @@ function makeClosureSignature(
   closureSignature.append(swift` in`);
   return closureSignature;
 }
+
+/**
+ * Takes a proposed name and modifies it to be unique given a list of properties.
+ * @param proposedName The proposed name that shouldn't conflict with any property.
+ * @param properties A list of properties the name shouldn't conflict with.
+ * @returns A name based on `proposedName` that doesn't match any existing property name.
+ */
+function makeUniqueName(
+  proposedName: string,
+  properties: { propertyName: string }[]
+): string {
+  // Assume conflicts are very rare and property lists are short, and just do a linear search. If
+  // we find a conflict, start over with the modified name.
+  for (let name = proposedName; ; name += "_") {
+    if (properties.every(prop => prop.propertyName != name)) {
+      return name;
+    }
+  }
+}
+
+/**
+ * Converts a value from "underscore_case" to "camelCase".
+ *
+ * This preserves any leading/trailing underscores.
+ */
+function camelCase(value: string): string {
```

Comment:
So it seems like the goal here is to have different behavior (ie, preserve leading and trailing underscores) than the `change-case` package - why not make that clear by using different function names rather than essentially overriding the functions? 

Alternately, would it be preferable to add a `preserveLeadingAndTrailingUnderscores` parameter available on the underlying functions so other packages could make use of it if needed?

Reply:
The underlying functions come from an external dependency, which is a weird constellation of dependencies for different modifications. I was originally going to try and submit a change upstream but it turns out the dependencies all work by first converting the input into space-separated words, e.g. `fooBarBaz` turns into `foo bar baz`, and I wasn’t sure if I could make the case for `_fooBar` to become `_foo bar`. Personally I think it would make sense. I also didn’t want to block on the dependency making any changes.

I could rename this but this new function is how I think `camelCase` _should_ work so I’d rather just hide the original and use this.

That said, I don’t know if this new function would be useful outside of `apollo-codegen-swift`. If other packages import the same `camelCase` dependency maybe they’d want this new behavior too? I figured we can get it into swift first and then consider moving it into a shared component if other components want it.

Reply:
Fair enough on not wanting to futz with the upstream dependency. 

I'm marginally in agreement about "This is how it should work" but I think if we're making changes to what happens under the hood, it should have a different name to make that clear.

Reply:
It should only have a different name if there’s a common agreement that something called `camelCase` should turn `_foo` into `foo`. I was quite surprised to find that was the existing behavior.

Reply:
I disagree with that - I think it should have a different name if expected behavior is different than the imported lib's expected behavior, even if you think the imported lib is wrong (unless you've made a PR to fix the imported lib's behavior). 

Again, I agree with your decision to not bother messing with the imported lib, but I think we should at least be clear that the expected behavior is different by explicitly naming our methods differently.

Reply:
The imported lib is just an implementation detail though. Callers of this function shouldn’t even care that the function relies on an imported lib to work. I’m tempted to just remove the imported lib entirely except I don’t want to risk introducing any behavioral changes (e.g. I think `camelCase(“FOO_bar”)` will turn into `”fooBar”`, which I don’t actually agree with, but I don’t want to make that kind of behavioral change without intent).

Reply:
IMO if it's the same name as the thing being imported, it quacks like an override to me 🦆.

Also, while the comments state that this preserves leading/trailing underscores, but it doesn't state that this is specifically why we're doing this "override". For future maintainers I'd prefer this have a different name for the extra clarity.

Reply:
I still don't understand how changing the name of this adds clarity. The "change-case" package does not provide a canonical definition of going between snake_case and camelCase, it's merely an implementation. The fact that our `camelCase` uses the `change-case` version internally is literally just an implementation detail, one that I wouldn't mind removing in the future since it seems like an unnecessary dependency. Change the name of this to `camelCasePreservingLeadingAndTrailingUnderscores`¹ implies the presence of an alternative `camelCase` function that literally isn't in scope (the imported function is now called `_camelCase`).

¹I really have no idea what you're suggesting we change it to anyway.

Reply:
I guess to make my position more clear, since I feel like we're going in circles: To be an "override" it has to be overriding something that a naïve reader would expect. For example, if I'm overriding something from the JS standard library, or if I'm using a well-known framework but overriding some of its behavior, that's an override. But in this case the fact that we're using the `change-case` package isn't something a naïve reader would expect and therefore we aren't "overriding" it, it's merely just an implementation detail.